!snapformat doesn't work in OpenJ9 jdmpview

> !snapformat Snap.trc.fmt.txt
Error executing DDR command !snapformat Snap.trc.fmt.txt : null
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.CombinedContext.executeDDRInteractiveCommand(CombinedContext.java:260)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.CombinedContext.execute(CombinedContext.java:170)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.Session.execute(Session.java:819)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.Session.execute(Session.java:773)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.tools.ToolsRegistry.executeJdmpviewCommand(ToolsRegistry.java:183)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.tools.ToolsRegistry.execute(ToolsRegistry.java:107)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.tools.ToolsRegistry.execute(ToolsRegistry.java:88)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.tools.ToolsRegistry.recordAndExecute(ToolsRegistry.java:72)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.Session.runInteractive(Session.java:747)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.Session.run(Session.java:671)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.DTFJView.launch(DTFJView.java:52)
	at openj9.dtfjview/com.ibm.jvm.dtfjview.DTFJView.main(DTFJView.java:46)
Caused by: java.lang.NoClassDefFoundError: com.ibm.jvm.trace.format.api.TracePoint
	at java.base/java.lang.Class.forNameImpl(Native Method)
	at java.base/java.lang.Class.forName(Class.java:339)
	at com.ibm.j9ddr.tools.ddrinteractive.commands.SnapFormatWrapperCommand.run(SnapFormatWrapperCommand.java:67)
	at com.ibm.j9ddr.tools.ddrinteractive.Context.tryCommand(Context.java:229)
	at com.ibm.j9ddr.tools.ddrinteractive.Context.execute(Context.java:202)
	at com.ibm.j9ddr.tools.ddrinteractive.DDRInteractive.execute(DDRInteractive.java:385)
	at com.ibm.j9ddr.command.CommandReader.processLine(CommandReader.java:79)
	at com.ibm.j9ddr.tools.ddrinteractive.DDRInteractive.processLine(DDRInteractive.java:360)
	... 16 more
Caused by: java.lang.ClassNotFoundException: com.ibm.jvm.trace.format.api.TracePoint
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:605)
	at java.base/java.lang.ClassLoader.loadClassHelper(ClassLoader.java:1134)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1049)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1032)
	... 24 more

tested JVM information:
openjdk version "11.0.1" 2018-10-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.1+13-201812091910)
Eclipse OpenJ9 VM AdoptOpenJDK (build master-842f12f84, JRE 11 Linux ppc64le-64-Bit Compressed References 20181209_47 (JIT enabled, AOT enabled)
OpenJ9   - 842f12f84
OMR      - 9b18ac96
JCL      - 14513b61ab based on jdk-11.0.1+13)

https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk11u-2018-12-09-21-49/OpenJDK11U-jdk_ppc64le_linux_openj9_2018-12-09-21-49.tar.gz

--- a/debugtools/DDR_VM/src/com/ibm/j9ddr/tools/ddrinteractive/commands/SnapFormatCommand.java
+++ b/debugtools/DDR_VM/src/com/ibm/j9ddr/tools/ddrinteractive/commands/SnapFormatCommand.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2013, 2018 IBM Corp. and others
+ * Copyright (c) 2013, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -31,6 +31,7 @@ import java.io.PrintStream;
 import java.nio.BufferUnderflowException;
 import java.util.ArrayList;
 import java.util.Iterator;
+import java.util.List;
 import java.util.StringTokenizer;
 
 import com.ibm.j9ddr.CorruptDataException;
@@ -42,32 +43,29 @@ import com.ibm.jvm.trace.format.api.TracePoint;
 import com.ibm.jvm.trace.format.api.TracePointImpl;
 import com.ibm.jvm.trace.format.api.TraceThread;
 
-public class SnapFormatCommand extends SnapBaseCommand 
-{
-	private TraceContext traceContext = null;
-	private InputStream[] messageFiles = new InputStream[2];
-	private boolean outputHeader = true;
-	
-	private static final String[] MESSAGEFILENAMES = {"TraceFormat.dat","J9TraceFormat.dat"};
-	private static final String[] DEFAULTMESSAGEFILEPATHS = new String[] {System.getProperty("user.dir"), System.getProperty("java.home") + File.separator + "lib"} ;
-	private static final PrintStream dummyOut = new PrintStream(new OutputStream() {
-		@Override
-		public void write(int b) throws IOException 
-		{
-			// do nothing
-		}
-	});
-	
-	private static abstract class TraceFilterExpression
-	{
-		public abstract boolean matches(TracePoint tp);
-		
-		public static TraceFilterExpression parseExpression(String expression)
-		{
-			TraceFilterExpression result = null;
+public class SnapFormatCommand extends SnapBaseCommand {
 
+	private TraceContext traceContext;
+	private final List<InputStream> messageFiles;
+	private boolean outputHeader;
+
+	private static final String[] MESSAGEFILENAMES = { "TraceFormat.dat", "J9TraceFormat.dat", "OMRTraceFormat.dat" };
+	private static final String[] DEFAULTMESSAGEFILEPATHS = {
+			System.getProperty("user.dir"),
+			System.getProperty("java.home") + File.separator + "lib"
+	};
+
+	private static abstract class TraceFilterExpression {
+
+		TraceFilterExpression() {
+			super();
+		}
+
+		public abstract boolean matches(TracePoint tp);
+
+		public static TraceFilterExpression parseExpression(String expression) {
 			StringTokenizer toker = new StringTokenizer(expression, "()&|!,", true);
-			ArrayList<String> tokens = new ArrayList<String>();
+			List<String> tokens = new ArrayList<>();
 			while (toker.hasMoreTokens()) {
 				String token = toker.nextToken().trim();
 				if (0 != token.length()) {
@@ -79,76 +77,56 @@ public class SnapFormatCommand extends SnapBaseCommand
 					}
 				}
 			}
-			result = parseExpression(tokens);
-			if (null == result) {
-				throw new IllegalArgumentException("Failed to parse filter expression: no result");
-			}
+			TraceFilterExpression result = parseExpression(tokens);
 			if (!tokens.isEmpty()) {
 				throw new IllegalArgumentException("Failed to parse filter expression: unparsed text");
 			}
 			return result;
 		}
 
-		private static TraceFilterExpression parseExpression(ArrayList<String> tokens)
-		{
-			TraceFilterExpression result = null;
+		private static TraceFilterExpression parseExpression(List<String> tokens) {
+			TraceFilterExpression result = parseTerm(tokens);
 
-			result = parseTerm(tokens);
 			while (!tokens.isEmpty()) {
 				String token = tokens.get(0);
 				if (token.equals("|") || token.equals(",")) {
 					tokens.remove(0);
 					TraceFilterExpression lhs = result;
 					TraceFilterExpression rhs = parseTerm(tokens);
-					if (null == rhs) {
-						throw new IllegalArgumentException("Failed to parse filter expression: missing operand to OR");
-					}
 					result = new TraceFilterOrExpression(lhs, rhs);
 				} else {
 					break;
 				}
 			}
-			
-			if (null == result) { 
-				throw new IllegalArgumentException("Failed to parse filter expression: expected expression");
-			}
+
 			return result;
 		}
-		
-		private static TraceFilterExpression parseTerm(ArrayList<String> tokens)
-		{
-			TraceFilterExpression result = null;
 
-			result = parsePrimary(tokens);
+		private static TraceFilterExpression parseTerm(List<String> tokens) {
+			TraceFilterExpression result = parsePrimary(tokens);
+
 			while (!tokens.isEmpty()) {
 				String token = tokens.get(0);
 				if (token.equals("&")) {
 					tokens.remove(0);
 					TraceFilterExpression lhs = result;
 					TraceFilterExpression rhs = parsePrimary(tokens);
-					if (null == rhs) {
-						throw new IllegalArgumentException("Failed to parse filter expression: missing operand to AND");
-					}
 					result = new TraceFilterAndExpression(lhs, rhs);
 				} else {
 					break;
 				}
 			}
-			
-			if (null == result) { 
-				throw new IllegalArgumentException("Failed to parse filter expression: expected expression");
-			}
+
 			return result;
 		}
-		
-		private static TraceFilterExpression parsePrimary(ArrayList<String> tokens)
-		{
+
+		private static TraceFilterExpression parsePrimary(List<String> tokens) {
 			TraceFilterExpression result = null;
 
-			if (tokens.isEmpty()) { 
+			if (tokens.isEmpty()) {
 				throw new IllegalArgumentException("Failed to parse filter expression: expected expression");
 			}
-			
+
 			String token = tokens.get(0);
 			if (token.equals("(")) {
 				// subexpression
@@ -158,19 +136,13 @@ public class SnapFormatCommand extends SnapBaseCommand
 					throw new IllegalArgumentException("Failed to parse filter expression: expected closing parenthesis");
 				}
 				tokens.remove(0);
-
 			} else if (token.equals("!")) {
 				// not expression
 				tokens.remove(0);
 				result = parsePrimary(tokens);
-				if (null == result) {
-					throw new IllegalArgumentException("Failed to parse filter expression: missing operand to NOT");
-				}
 				result = new TraceFilterNotExpression(result);
-				
 			} else if (token.equals(")") || token.equals("&") || token.equals("|") || token.equals(",")) {
 				throw new IllegalArgumentException("Failed to parse filter expression: expected expression, found " + token);
-				
 			} else {
 				// anything else
 				tokens.remove(0);
@@ -180,91 +152,87 @@ public class SnapFormatCommand extends SnapBaseCommand
 			return result;
 		}
 	}
-	
-	private static class TraceFilterOrExpression extends TraceFilterExpression
-	{
-		private TraceFilterExpression lhs;
-		private TraceFilterExpression rhs;
 
-		public TraceFilterOrExpression(TraceFilterExpression lhs, TraceFilterExpression rhs)
-		{
+	private static final class TraceFilterOrExpression extends TraceFilterExpression {
+
+		private final TraceFilterExpression lhs;
+		private final TraceFilterExpression rhs;
+
+		public TraceFilterOrExpression(TraceFilterExpression lhs, TraceFilterExpression rhs) {
 			this.lhs = lhs;
 			this.rhs = rhs;
 		}
-		
-		public boolean matches(TracePoint tp)
-		{
+
+		@Override
+		public boolean matches(TracePoint tp) {
 			return lhs.matches(tp) || rhs.matches(tp);
 		}
-		
-		public String toString()
-		{
+
+		@Override
+		public String toString() {
 			return "(" + lhs + "|" + rhs + ")";
 		}
 	}
 
-	private static class TraceFilterAndExpression extends TraceFilterExpression
-	{
-		private TraceFilterExpression lhs;
-		private TraceFilterExpression rhs;
-		
-		public TraceFilterAndExpression(TraceFilterExpression lhs, TraceFilterExpression rhs)
-		{
+	private static final class TraceFilterAndExpression extends TraceFilterExpression {
+
+		private final TraceFilterExpression lhs;
+		private final TraceFilterExpression rhs;
+
+		public TraceFilterAndExpression(TraceFilterExpression lhs, TraceFilterExpression rhs) {
 			this.lhs = lhs;
 			this.rhs = rhs;
 		}
-		
-		public boolean matches(TracePoint tp)
-		{
+
+		@Override
+		public boolean matches(TracePoint tp) {
 			if (lhs.matches(tp)) {
 				return rhs.matches(tp);
 			}
 			return false;
 		}
-		
-		public String toString()
-		{
+
+		@Override
+		public String toString() {
 			return "(" + lhs + "&" + rhs + ")";
 		}
 	}
 
-	private static class TraceFilterNotExpression extends TraceFilterExpression
-	{
-		protected TraceFilterExpression expression;
-		
-		public TraceFilterNotExpression(TraceFilterExpression expression)
-		{
+	private static final class TraceFilterNotExpression extends TraceFilterExpression {
+
+		private final TraceFilterExpression expression;
+
+		public TraceFilterNotExpression(TraceFilterExpression expression) {
 			this.expression = expression;
 		}
-		
-		public boolean matches(TracePoint tp)
-		{
+
+		@Override
+		public boolean matches(TracePoint tp) {
 			return !expression.matches(tp);
 		}
-		
-		public String toString()
-		{
+
+		@Override
+		public String toString() {
 			return "!" + expression;
 		}
 	}
 
-	private static class TraceFilter extends TraceFilterExpression
-	{
+	private static final class TraceFilter extends TraceFilterExpression {
+
 		private String component = null;
 		private String type = null;
 		private int idLow = -1;
 		private int idHigh = -1;
 
 		/* Format of a tracepoint specification:
-		 * 	<component>[{<type>}]
+		 *  <component>[{<type>}]
 		 * or
-		 * 	<component>.<id>
+		 *  <component>.<id>
 		 * or
-		 * 	<component>.<id-range>
+		 *  <component>.<id-range>
 		 */
-		public TraceFilter(String spec)
-		{
-			int index = 0;	
+		public TraceFilter(String spec) {
+			int index = 0;
 			int typeIndex = spec.indexOf('{', index);
 			if (-1 == typeIndex) {
 				// no type, maybe an id?
@@ -286,15 +254,15 @@ public class SnapFormatCommand extends SnapBaseCommand
 				component = spec.substring(index, typeIndex);
 				type = spec.substring(typeIndex + 1, spec.length() - 1);
 			}
-			
+
 			if ((null != component && ((0 == component.length()) || component.equalsIgnoreCase("all")))) {
 				// "all" is the default
 				component = null;
 			}
 		}
-		
-		public boolean matches(TracePoint tp)
-		{
+
+		@Override
+		public boolean matches(TracePoint tp) {
 			boolean result = true;
 			if (null != component) {
 				if (component.equals(tp.getComponent())) {
@@ -328,9 +296,9 @@ public class SnapFormatCommand extends SnapBaseCommand
 
 			return result;
 		}
-		
-		public String toString()
-		{
+
+		@Override
+		public String toString() {
 			StringBuilder builder = new StringBuilder();
 			if (null != component) {
 				builder.append(component);
@@ -353,140 +321,163 @@ public class SnapFormatCommand extends SnapBaseCommand
 			return builder.toString();
 		}
 	}
-	
-	
-	public SnapFormatCommand()
-	{
-	}	
+
+	public SnapFormatCommand() {
+		super();
+		traceContext = null;
+		messageFiles = new ArrayList<>();
+		outputHeader = true;
+	}
 
 	/**
 	 * Options should be:
-	 * 	-f ouputFile
-	 * 	-t vmthread
-	 * 	-d .dat file path
+	 *  -f ouputFile
+	 *  -t vmthread
+	 *  -d .dat file path
 	 */
 	public void run(String command, String[] args, Context context,
 			PrintStream out) throws DDRInteractiveCommandException {
-		
 		String fileName = null;
 		String userDatPath = null;
 		String userThreadId = null;
 		String messageFilePath = null;
+		List<String> messageFileStrings = new ArrayList<>();
 		TraceFilterExpression specFilter = null;
-		
+
 		/* Usage:
-		 * 	!snapformat [<filename>]
-		 * 		- format trace buffers to a specified file or stdout
-		 * 	!snapformat [-f <filename>] [-d <datfile_directory>] [-t <j9vmthread id>] 
-		 * 		- format trace buffers for all threads or just the specified thread to a file or
-		 * 		  stdout using the .dat files in the specified dir
-		 * 
+		 *  !snapformat [<filename>]
+		 *      - format trace buffers to a specified file or stdout
+		 *  !snapformat [-f <filename>] [-d <datfile_directory>] [-t <j9vmthread id>]
+		 *      - format trace buffers for all threads or just the specified thread to a file or
+		 *        stdout using the .dat files in the specified dir
 		 */
-		
-		if (args.length == 0 ){
+
+		if (args.length == 0) {
 			fileName = null; // Write formatted trace to the console.
-		} else if( args.length == 1 ) {
+		} else if (args.length == 1) {
 			fileName = args[0];
 		} else {
 			/* Walk the arguments and look for flag/value pairs. */
-			for( int i = 0; i < args.length - 1; ) {
+			for (int i = 0; i < args.length - 1;) {
 				String flag = args[i++];
 				String value = args[i++];
-				if( "-f".equals(flag)) {
+				if ("-f".equals(flag)) {
 					fileName = value;
-				} else if( "-d".equals(flag)) {
+				} else if ("-d".equals(flag)) {
 					userDatPath = value;
-				} else if( "-t".equals(flag)) {
+				} else if ("-t".equals(flag)) {
 					userThreadId = value;
 					outputHeader = false;
-				} else if( "-s".equals(flag)) {
+				} else if ("-s".equals(flag)) {
 					specFilter = TraceFilterExpression.parseExpression(value);
 					outputHeader = false;
 				}
 			}
 		}
-	
+
 		/* Look for message files, search order is:
-		 * - User specified location to this command.
+		 * - user specified location to this command
 		 * - as resources located under same class loader as TraceContext
 		 * - current working directory
-		 * - jre/lib for the current jre 
+		 * - jre/lib for the current jre
 		 */
-		boolean foundDatFiles = false;
-		if( userDatPath != null ) {
-			int i = 0;
-			for( String name : MESSAGEFILENAMES ) {
-				File f = new File(userDatPath + File.separator + name);
-				if( !f.exists() ) {
-					out.printf("Error locating .dat file %s on user path %s\n", name, userDatPath);
-					return;
-				} else {
+		if (userDatPath != null) {
+			messageFilePath = userDatPath;
+			for (String name : MESSAGEFILENAMES) {
+				File f = new File(userDatPath, name);
+				if (f.exists()) {
 					try {
-						messageFiles[i++] = new FileInputStream(f);
-						foundDatFiles = true;
+						messageFiles.add(new FileInputStream(f));
+						messageFileStrings.add(name);
 					} catch (FileNotFoundException e) {
 						// We've done f.exists(), this should be fine.
-						// foundDatFiles will be false if not.
 					}
 				}
 			}
 		} else {
-			int i = 0;
-			String[] messageFileStrings = new String[MESSAGEFILENAMES.length];
-			for( String name : MESSAGEFILENAMES ) {
+			for (String name : MESSAGEFILENAMES) {
 				InputStream s = TraceContext.class.getResourceAsStream('/' + name);
-				if ( null != s ) {
-					messageFileStrings[i] = TraceContext.class.getResource('/' + name).toString();
-					messageFiles[i++] = s;
-					foundDatFiles = true;
+				if (s != null) {
+					messageFiles.add(s);
+					messageFileStrings.add(TraceContext.class.getResource('/' + name).toString());
 				}
 			}
-			if( foundDatFiles ) {
-				out.printf("Formatting trace using format dat files from %s and %s\n", messageFileStrings[0], messageFileStrings[1]);
-			} else {
-				for( String path: DEFAULTMESSAGEFILEPATHS ) {
-					for( String name : MESSAGEFILENAMES ) {
-						File f = new File(path + File.separator + name);
-						if( f.exists() ) {
+			if (messageFiles.isEmpty()) {
+				for (String path : DEFAULTMESSAGEFILEPATHS) {
+					for (String name : MESSAGEFILENAMES) {
+						File f = new File(path, name);
+						if (f.exists()) {
 							try {
-								messageFiles[i++] = new FileInputStream(f);
-								foundDatFiles = true;
+								messageFiles.add(new FileInputStream(f));
+								messageFileStrings.add(name);
 							} catch (FileNotFoundException e) {
 								// We've done f.exists(), this should be fine.
-								// foundDatFiles will be false if not.
 							}
 						}
 					}
-					if( foundDatFiles ) {
+
+					if (!messageFiles.isEmpty()) {
 						messageFilePath = path;
-						out.printf("Formatting trace using format dat files from %s and %s from %s\n", MESSAGEFILENAMES[0], MESSAGEFILENAMES[1], messageFilePath);
 						break;
 					}
 				}
 			}
-			if( !foundDatFiles ) {
-				out.printf("Unable to find %s and %s in %s or %s\n", MESSAGEFILENAMES[0], MESSAGEFILENAMES[1], DEFAULTMESSAGEFILEPATHS[0], DEFAULTMESSAGEFILEPATHS[1]);
-				return;
+		}
+
+		if (messageFiles.isEmpty()) {
+			String leader = "Unable to find any of";
+
+			for (String name : MESSAGEFILENAMES) {
+				out.printf("%s %s", leader, name);
+				leader = ",";
+			}
+
+			if (messageFilePath != null) {
+				out.printf(" in %s%n", messageFilePath);
+			} else {
+				out.printf(" in %s or %s%n", DEFAULTMESSAGEFILEPATHS[0], DEFAULTMESSAGEFILEPATHS[1]);
+			}
+
+			return;
+		} else {
+			String leader = "Formatting trace using format dat files";
+
+			for (String name : messageFileStrings) {
+				out.printf("%s %s", leader, name);
+				leader = ",";
+			}
+
+			if (messageFilePath != null) {
+				out.printf(" from %s%n", messageFilePath);
+			} else {
+				out.println();
 			}
 		}
-		
+
 		if (outputHeader) {
 			extractTraceData(context, out);
 		} else {
+			final PrintStream dummyOut = new PrintStream(new OutputStream() {
+				@Override
+				public void write(int octet) {
+					// do nothing
+				}
+			});
+
 			extractTraceData(context, dummyOut);
 		}
-		
-		if( traceContext == null ) {
+
+		if (traceContext == null) {
 			out.println("Unable to create trace context, command failed.");
 			return;
 		}
-		
+
 		long threadId = 0;
-		if( userThreadId != null ) {
-			boolean is64BitPlatform = (context.process.bytesPerPointer() == 8) ? true : false;
+		if (userThreadId != null) {
+			boolean is64BitPlatform = context.process.bytesPerPointer() == 8;
 			threadId = CommandUtils.parsePointer(userThreadId, is64BitPlatform);
 		}
-		
+
 		/* Create the stream to write the trace to.
 		 * The specified output file or "out" if we
 		 * are writing to the console.
@@ -495,7 +486,7 @@ public class SnapFormatCommand extends SnapBaseCommand
 		 */
 		PrintStream traceOut = out;
 		PrintStream filePrintStream = null;
-		if( fileName != null ) {
+		if (fileName != null) {
 			try {
 				filePrintStream = new PrintStream(fileName);
 				traceOut = filePrintStream;
@@ -503,43 +494,42 @@ public class SnapFormatCommand extends SnapBaseCommand
 				out.printf("Unable to write formatted trace to file %s\n", fileName);
 			}
 		}
-	
+
 		if (outputHeader) {
 			traceOut.println(traceContext.summary());
 		}
-		
+
 		try {
 			Iterator<TracePoint> tpIterator = null;
-			if( userThreadId != null ) {
-				Iterator<TraceThread> threadsIterator = (Iterator<TraceThread>)traceContext.getThreads();
+			if (userThreadId != null) {
+				Iterator<TraceThread> threadsIterator = (Iterator<TraceThread>) traceContext.getThreads();
 				boolean foundThread = false;
 				while (threadsIterator.hasNext()) {
 					TraceThread thread = threadsIterator.next();
-					if( thread.getThreadID() == threadId ) {
+					if (thread.getThreadID() == threadId) {
 						foundThread = true;
-						tpIterator = (Iterator<TracePoint>)thread.getIterator();
+						tpIterator = (Iterator<TracePoint>) thread.getIterator();
 					}
 				}
-				if( !foundThread ) {
+				if (!foundThread) {
 					out.printf("Unable to find thread %s in trace data\n", userThreadId);
 				}
 			} else {
-				tpIterator = (Iterator<TracePoint>)traceContext.getTracepoints();	
+				tpIterator = (Iterator<TracePoint>) traceContext.getTracepoints();
 			}
-			
-			if( tpIterator != null ) {
+
+			if (tpIterator != null) {
 				int tpCount = printTracePoints(traceOut, tpIterator, specFilter);
-				out.printf("Completed processing of %d tracepoints with %d warnings and %d errors\n", traceContext.getTotalTracePoints(), traceContext.getWarningCount(), traceContext.getErrorCount() );
+				out.printf("Completed processing of %d tracepoints with %d warnings and %d errors\n", traceContext.getTotalTracePoints(), traceContext.getWarningCount(), traceContext.getErrorCount());
 			}
 		} catch (Exception e) {
 			e.printStackTrace();
 		}
-		
-		if( filePrintStream != null ) {
+
+		if (filePrintStream != null) {
 			out.println("Snap trace written to: " + fileName);
 			filePrintStream.close();
 		}
-		
 	}
 
 	private int printTracePoints(PrintStream traceOut, Iterator<TracePoint> tpIterator, TraceFilterExpression specFilter) {
@@ -548,7 +538,7 @@ public class SnapFormatCommand extends SnapBaseCommand
 		while (tpIterator.hasNext()) {
 			TracePoint tp = tpIterator.next();
 			TracePointImpl tracepoint = null;
-			if( tp instanceof TracePointImpl ) {
+			if (tp instanceof TracePointImpl) {
 				tracepoint = (TracePointImpl) tp;
 			}
 			if ((tracepoint != null) && ((specFilter == null) || specFilter.matches(tracepoint))) {
@@ -562,19 +552,18 @@ public class SnapFormatCommand extends SnapBaseCommand
 				try {
 					parameters = tracepoint.getFormattedParameters();
 					if (parameters == null || parameters.length() == 0) {
-					traceContext.error(traceContext, "null parameter data for trace point "+component+"."+tpID);
+						traceContext.error(traceContext, "null parameter data for trace point " + component + "." + tpID);
 					}
 				} catch (BufferUnderflowException e) {
 					/* This may be thrown, but there's essentially nothing we can do about it at this level so
 					 * just report it
 					 */
-					traceContext.error(traceContext, "Underflow accessing parameter data for trace point "+component+"."+tpID);
+					traceContext.error(traceContext, "Underflow accessing parameter data for trace point " + component + "." + tpID);
 				}
-				
+
 				StringBuilder formatted = new StringBuilder();
-				
+
 				formatted.append(tracepoint.getFormattedTime());
-				
 
 				/* append thread id */
 				formatted.append(" ").append((current != thread ? "*" : " "));
@@ -584,7 +573,7 @@ public class SnapFormatCommand extends SnapBaseCommand
 				/* append component and padding */
 				StringBuilder fullTracepointID = new StringBuilder(component);
 				fullTracepointID.append(container != null ? "(" + container + ")" : "").append(".").append(tpID);
-				formatted.append(String.format("%-20s",fullTracepointID.toString()));
+				formatted.append(String.format("%-20s", fullTracepointID.toString()));
 				formatted.append(" ");
 
 				formatted.append(tracepoint.getType());
@@ -597,30 +586,28 @@ public class SnapFormatCommand extends SnapBaseCommand
 		}
 		return tpCount;
 	}
-	
-	@Override
-	protected void writeHeaderBytesToTrace(Context context, byte[] headerBytes, PrintStream out) 
-	{
 
+	@Override
+	protected void writeHeaderBytesToTrace(Context context, byte[] headerBytes, PrintStream out) {
 		try {
-			traceContext = TraceContext.getContext(headerBytes, headerBytes.length, messageFiles[0] );
-			traceContext.addMessageData(messageFiles[1]);
+			traceContext = TraceContext.getContext(headerBytes, headerBytes.length, messageFiles.get(0));
+
+			for (int i = 1, n = messageFiles.size(); i < n; ++i) {
+				traceContext.addMessageData(messageFiles.get(i));
+			}
 		} catch (IOException e) {
 			// TODO Auto-generated catch block
 			e.printStackTrace();
 		}
-
 	}
 
 	@Override
-	protected void writeBytesToTrace(Context context, long address,
-			int bufferSize, PrintStream out) {
-		
-		if( traceContext == null ) {
+	protected void writeBytesToTrace(Context context, long address, int bufferSize, PrintStream out) {
+		if (traceContext == null) {
 			out.println("Error - trace buffer passed before context created.");
 			return;
 		}
-		
+
 		byte[] data = new byte[bufferSize];
 		try {
 			context.process.getBytesAt(address, data);
@@ -632,7 +619,6 @@ public class SnapFormatCommand extends SnapBaseCommand
 			out.println("Problem reading " + bufferSize + " bytes from 0x" +  Long.toHexString(address) + ". Trace file may contain partial or damaged data.");
 		}
 		traceContext.addData(data);
-
 	}
 
 }

--- a/debugtools/DDR_VM/src/com/ibm/j9ddr/tools/ddrinteractive/commands/SnapFormatWrapperCommand.java
+++ b/debugtools/DDR_VM/src/com/ibm/j9ddr/tools/ddrinteractive/commands/SnapFormatWrapperCommand.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2013, 2015 IBM Corp. and others
+ * Copyright (c) 2013, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -38,11 +38,10 @@ public class SnapFormatWrapperCommand extends Command {
 		addCommand("snapformat", "[<filename>]", "format trace buffers to a specified file or stdout");
 		addCommand("snapformat", "-help", "print detailed help");
 		addCommand("snapformat", "[-f <filename>] [-d <datfile_directory>] [-t <j9vmthread id>] [-s <filter>]", "format trace buffers for all threads or just the specified thread to a file or stdout using the specified .dat files");
-	}	
+	}
 
 	public void run(String command, String[] args, Context context,
 			PrintStream out) throws DDRInteractiveCommandException {
-		
 		if ((args.length > 0) && (args[0].equals("-help") || args[0].equals("-h") || args[0].equals("-?"))) {
 			out.println("Format selected trace buffers.");
 			out.println();
@@ -64,35 +63,21 @@ public class SnapFormatWrapperCommand extends Command {
 			return;
 		}
 		try {
-			Class snapFormatCommandCls = Class.forName("com.ibm.j9ddr.tools.ddrinteractive.commands.SnapFormatCommand");
+			Class<?> snapFormatCommandCls = Class.forName("com.ibm.j9ddr.tools.ddrinteractive.commands.SnapFormatCommand");
 			Object snapFormatCommandInstance = snapFormatCommandCls.newInstance();
-			if( snapFormatCommandInstance instanceof Command) {
-				Command c = (Command)snapFormatCommandInstance;
+			if (snapFormatCommandInstance instanceof Command) {
+				Command c = (Command) snapFormatCommandInstance;
 				c.run(command, args, context, out);
 			} else {
 				throw new DDRInteractiveCommandException("Unable to format trace. Could not create formatter.");
 			}
-		} catch (ClassNotFoundException e) {
-			//e.printStackTrace(out);
-			throw new DDRInteractiveCommandException("Unable to format trace. " + e.getMessage(), e);
-		} catch (IllegalAccessException e) {
-			//e.printStackTrace(out);
-			throw new DDRInteractiveCommandException("Unable to format trace. " + e.getMessage(), e);
-		} catch (InstantiationException e) {
-			//e.printStackTrace(out);
-			throw new DDRInteractiveCommandException("Unable to format trace. " + e.getMessage(), e);
-		} catch (SecurityException e) {
-			//e.printStackTrace(out);
-			throw new DDRInteractiveCommandException("Unable to format trace. " + e.getMessage(), e);
-		} catch (IllegalArgumentException e) {
-			//e.printStackTrace(out);
+		} catch (ClassNotFoundException
+				| IllegalAccessException
+				| IllegalArgumentException
+				| InstantiationException
+				| SecurityException e) {
 			throw new DDRInteractiveCommandException("Unable to format trace. " + e.getMessage(), e);
 		}
-//		catch (java.lang.Error e) {
-//			/* Naughty catching this but good for catching unresolved com.ibm.jvm.trace.* problems. */
-//			//e.printStackTrace(out);
-//			throw new DDRInteractiveCommandException("Unable to format trace. " + e.getMessage(), e);
-//		}
 	}
 
 }

--- a/jcl/src/openj9.dtfj/share/classes/com/ibm/dtfj/image/j9/ImageFactory.java
+++ b/jcl/src/openj9.dtfj/share/classes/com/ibm/dtfj/image/j9/ImageFactory.java
@@ -1,6 +1,6 @@
 /*[INCLUDE-IF Sidecar18-SE]*/
 /*******************************************************************************
- * Copyright (c) 2004, 2018 IBM Corp. and others
+ * Copyright (c) 2004, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -66,6 +66,7 @@ public class ImageFactory implements com.ibm.dtfj.image.ImageFactory {
 	private ClassLoader imageFactoryClassLoader;
 	private File tmpdir = null; // the directory which holds any extracted files
 
+	/*[IF Sidecar19-SE]*/
 	static {
 		/* 
 		 * Even though j9ddr.jar is a resource within the module openj9.dtfj,
@@ -99,9 +100,10 @@ public class ImageFactory implements com.ibm.dtfj.image.ImageFactory {
 			exportToAll.invoke(null, thisModule, "com.ibm.java.diagnostics.utils"); //$NON-NLS-1$
 			exportToAll.invoke(null, thisModule, "com.ibm.java.diagnostics.utils.commands"); //$NON-NLS-1$
 		} catch (Exception e) {
-			// assume pre-Java 9 jvm
+			throw new InternalError("Failed to adjust module exports", e); //$NON-NLS-1$
 		}
 	}
+	/*[ENDIF] Sidecar19-SE*/
 	
 	/**
 	 * This public constructor is intended for use with Class.newInstance().

--- a/jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java
+++ b/jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java
@@ -1,6 +1,6 @@
 /*[INCLUDE-IF Sidecar18-SE]*/
 /*******************************************************************************
- * Copyright (c) 2004, 2017 IBM Corp. and others
+ * Copyright (c) 2004, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -207,7 +207,6 @@ public class Main {
 			}
 			return result;
 		}
-
 
 		public Object buildSymbol(Object addressSpace, String functionName, long relocatedFunctionAddress) {
 			return new Object();
@@ -474,25 +473,30 @@ public class Main {
 		for (Iterator iter = _dump.getAdditionalFileNames(); iter.hasNext();) {
 			files.add(iter.next());
 		}
-		
-			
+
 		try {
-			// Add the trace formatting template files (Traceformat.dat and J9TraceFormat.dat) into the zip
+			// Add trace formatting template files (Traceformat.dat, J9TraceFormat.dat and OMRTraceFormat.dat) into the archive.
 			String lib_dir = System.getProperty("java.home") + File.separator + "lib" + File.separator;
 			String trace = lib_dir + "TraceFormat.dat";
-			if (new File(trace).exists())
+			if (new File(trace).exists()) {
 				files.add(trace);
+			}
 			String j9trace = lib_dir + "J9TraceFormat.dat";
-			if (new File(j9trace).exists())
+			if (new File(j9trace).exists()) {
 				files.add(j9trace);
-			
+			}
+			String omrtrace = lib_dir + "OMRTraceFormat.dat";
+			if (new File(omrtrace).exists()) {
+				files.add(omrtrace);
+			}
+
 			// Add the debugger extension library into the zip, available only on AIX
 			String osName = System.getProperty("os.name");
 			if (osName != null && osName.equalsIgnoreCase("AIX")) {
 				files.add("libdbx_j9.so");
 			}
 		} catch (Exception e) {
-				// Ignore
+			// Ignore
 		}
 
 		try {
@@ -635,7 +639,6 @@ public class Main {
 		}
 	}
 
-	
 	private native long getEnvironmentPointer(IAbstractAddressSpace dump) throws Exception;
 	private native void doCommand(IAbstractAddressSpace dump, String command) throws Exception;
 

--- a/jcl/src/openj9.dtfj/share/classes/module-info.java
+++ b/jcl/src/openj9.dtfj/share/classes/module-info.java
@@ -1,6 +1,6 @@
 /*[INCLUDE-IF Sidecar19-SE]*/
 /*******************************************************************************
- * Copyright (c) 2016, 2017 IBM Corp. and others
+ * Copyright (c) 2016, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -32,6 +32,7 @@ module openj9.dtfj {
 	requires transitive java.desktop;
 	requires transitive java.logging;
 	requires java.xml;
+	requires openj9.traceformat;
 	/*[IF PLATFORM-mz31 | PLATFORM-mz64]*/
 	requires com.ibm.jzos;
 	/*[ENDIF]*/

--- a/jcl/src/openj9.traceformat/share/classes/com/ibm/jvm/format/TraceFormat.java
+++ b/jcl/src/openj9.traceformat/share/classes/com/ibm/jvm/format/TraceFormat.java
@@ -1,6 +1,6 @@
 /*[INCLUDE-IF Sidecar18-SE]*/
 /*******************************************************************************
- * Copyright (c) 2000, 2018 IBM Corp. and others
+ * Copyright (c) 2000, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -448,6 +448,16 @@ final public class TraceFormat
 			TraceFormat.outStream.println("*** Could not find a J9TraceFormat.dat file. JVM trace points may be formatted incorrectly.");
 		}
 		
+		/*
+		 * Now load OMRTraceFormat.dat if available.
+		 */
+		String omrFormatFilePath = findDatFile(dirsToSearch, "OMRTraceFormat.dat");
+
+		if (omrFormatFilePath != null) {
+			/* we found a suitable dat file */
+			tryMessageFileInstantiation(omrFormatFilePath);
+		}
+
 		/*
 		 * Also load TraceFormat.dat as it contains the trace points for the class libraries.
 		 */

--- a/jcl/src/openj9.traceformat/share/classes/com/ibm/jvm/traceformat/TraceFormat.java
+++ b/jcl/src/openj9.traceformat/share/classes/com/ibm/jvm/traceformat/TraceFormat.java
@@ -1,12 +1,6 @@
 /*[INCLUDE-IF Sidecar19-SE]*/
-/*[IF Sidecar19-SE]*/
-package com.ibm.jvm.traceformat;
-/*[ELSE]
-package com.ibm.jvm;
-/*[ENDIF]*/
-
 /*******************************************************************************
- * Copyright (c) 2010, 2018 IBM Corp. and others
+ * Copyright (c) 2010, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -26,6 +20,11 @@ package com.ibm.jvm;
  *
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
  *******************************************************************************/
+/*[IF Sidecar19-SE]*/
+package com.ibm.jvm.traceformat;
+/*[ELSE]
+package com.ibm.jvm;
+/*[ENDIF]*/
 
 import java.io.File;
 import java.io.FileInputStream;
@@ -55,7 +54,7 @@ import com.ibm.jvm.trace.format.api.TraceThread;
 
 /**
  * !!! WARNING !!!
- * Adding any new top level classes in this file requires modification to jcl\jcl_build.mk
+ * Adding any new top level classes in this file requires modification to jcl/jcl_build.mk
  * to ensure the new classes are included within traceformat.jar for Java 8.
  */
 public class TraceFormat
@@ -528,9 +527,11 @@ class MessageFile extends ProgramOption {
 	List messageFiles = new LinkedList();
 
 	String getDescription() {
-		return "A comma separated list of files containing the trace format strings. By default the following files are used:"+System.getProperty("line.separator")+
-			   "		$JAVA_HOME/lib/J9TraceFormat.dat"+System.getProperty("line.separator")+
-			   "		$JAVA_HOME/lib/TraceFormat.dat";
+		String eol = System.getProperty("line.separator", "\n");
+		return "A comma separated list of files containing the trace format strings. By default the following files are used:" + eol
+			 + "  $JAVA_HOME/lib/J9TraceFormat.dat" + eol
+			 + "  $JAVA_HOME/lib/OMRTraceFormat.dat" + eol
+			 + "  $JAVA_HOME/lib/TraceFormat.dat";
 	}
 
 	String getName() {
@@ -554,13 +555,11 @@ class MessageFile extends ProgramOption {
 				token = st.nextToken();
 				/* construct files from the tokens. These files must exist */
 				File datFile = new File(token);
-				if( !datFile.exists() ) {
-					throw new FileNotFoundException(token);
+				if (!datFile.exists()) {
+					throw new IllegalArgumentException("dat file \"" + token + "\" not found");
 				}
 				messageFiles.add(datFile);
 			}
-		} catch (FileNotFoundException e) {
-			throw new IllegalArgumentException("dat file \""+token+"\" not found");
 		} catch (SecurityException e) {
 			throw new IllegalArgumentException("The application does not have permission to access the specified dat file, \""+token+"\"");
 		}
@@ -571,6 +570,7 @@ class MessageFile extends ProgramOption {
 		dir = dir.concat(File.separator).concat("lib").concat(File.separator);
 		
 		setValue(dir + "J9TraceFormat.dat");
+		setValue(dir + "OMRTraceFormat.dat");
 		try {
 			setValue(dir + "TraceFormat.dat");
 		} catch (IllegalArgumentException e) {

--- a/jcl/src/openj9.traceformat/share/classes/module-info.java
+++ b/jcl/src/openj9.traceformat/share/classes/module-info.java
@@ -1,6 +1,6 @@
 /*[INCLUDE-IF Sidecar19-SE]*/
 /*******************************************************************************
- * Copyright (c) 2016, 2017 IBM Corp. and others
+ * Copyright (c) 2016, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -20,5 +20,6 @@
  *
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
  *******************************************************************************/
- module openj9.traceformat {
- }
+module openj9.traceformat {
+	exports com.ibm.jvm.trace.format.api;
+}

